provider/core/src/data_provider.rs

```diff
@@ -95,11 +95,23 @@ pub struct DataResponseMetadata {
     pub data_langid: Option<LanguageIdentifier>,
 }
 
+/// Dummy trait that lets us `dyn Drop`
+///
+/// `dyn Drop` isn't legal (and doesn't make sense since `Drop` is not
+/// implement on all destructible types). However, all trait objects come with
+/// a destructor, so we can just use an empty trait to get a destructor object.
+pub(crate) trait ErasedDestructor<'data>: IsCovariant<'data> {}
+impl<'data, T: IsCovariant<'data>> ErasedDestructor<'data> for T {}
+
+/// All that Yoke needs from the Cart type is the destructor. We can
+/// use a trait object to handle that.
+pub(crate) type ErasedCart<'data> = Rc<dyn ErasedDestructor<'data> + 'data>;
```

Comment:
IMO it would be if it:

 - did not have the lifetime
 - did not have the `IsCovariant<'data>` bound

These are both due to tricky constraints from our side and they make it not as general. My original plan was for it to be in the yoke crate but as long as we have the `IsCovariant<'data>` bound it doesn't make that much sense.

Once we get rid of the lifetime then this will be possible, I don't want to do it right now because the current design is too specific to us. Even if we can get rid of the `IsCovariant` bound I'd be okay with moving this upstream because `ErasedCart<'static>` would always work for folks who don't want to care about the lifetime. Requiring `IsCovariant` makes this much less generally useful

Reply:
There are also multiple ownership patterns one may want here; from an `Rc` to a `Box`.

Reply:
Will file a followup for moving this out

Reply:
https://github.com/unicode-org/icu4x/issues/1284